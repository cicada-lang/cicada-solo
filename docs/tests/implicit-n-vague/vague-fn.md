---
section: Implicit & Vague
title: Vague Function
---

# List

```cicada
datatype List(E: Type) {
  null: List(E)
  cons(head: E, tail: List(E)): List(E)
}
```

# vague-fn insertion -- on implicit-fn

```cicada
check (
  vague A,
  vague B,
  vague C,
  implicit T,
  x,
) => cons(List.null, cons(List.null, cons(List.null, T))): (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  implicit T: Type,
  x: T,
) -> Pair(List(A), Pair(List(B), Pair(List(C), Type)))

check (
  vague A,
  vague B,
  implicit T,
  x,
) => cons(List.null, cons(List.null, cons(List.null, T))): (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  implicit T: Type,
  x: T,
) -> Pair(List(A), Pair(List(B), Pair(List(C), Type)))

check (
  vague A,
  implicit T,
  x,
) => cons(List.null, cons(List.null, cons(List.null, T))): (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  implicit T: Type,
  x: T,
) -> Pair(List(A), Pair(List(B), Pair(List(C), Type)))

check (
  implicit T,
  x,
) => cons(List.null, cons(List.null, cons(List.null, T))): (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  implicit T: Type,
  x: T,
) -> Pair(List(A), Pair(List(B), Pair(List(C), Type)))
```

# vague-fn insertion -- on sigma

```cicada
check (
  vague A,
  vague B,
  vague C,
  vague D,
) => cons(List.null, cons(List.null, cons(List.null, List.null))) : (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  vague D: Type,
) -> Pair(List(A), Pair(List(B), Pair(List(C), List(D))))

check (
  vague A,
  vague B,
  vague C,
) => cons(List.null, cons(List.null, cons(List.null, List.null))) : (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  vague D: Type,
) -> Pair(List(A), Pair(List(B), Pair(List(C), List(D))))

check (
  vague A,
  vague B,
) => cons(List.null, cons(List.null, cons(List.null, List.null))) : (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  vague D: Type,
) -> Pair(List(A), Pair(List(B), Pair(List(C), List(D))))

check (
  vague A,
) => cons(List.null, cons(List.null, cons(List.null, List.null))) : (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  vague D: Type,
) -> Pair(List(A), Pair(List(B), Pair(List(C), List(D))))

check cons(List.null, cons(List.null, cons(List.null, List.null))): (
  vague A: Type,
  vague B: Type,
  vague C: Type,
  vague D: Type,
) -> Pair(List(A), Pair(List(B), Pair(List(C), List(D))))
```
