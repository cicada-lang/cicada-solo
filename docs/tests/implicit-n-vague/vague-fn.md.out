(vague A, vague B, vague C, implicit T, x) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | T]]]: (vague A: Type, vague B: Type, vague C: Type, implicit T: Type, x: T) -> [_: List(A), _1: List(B), _2: List(C) | Type]
(vague A, vague B, vague C, implicit T, x) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | T]]]: (vague A: Type, vague B: Type, vague C: Type, implicit T: Type, x: T) -> [_: List(A), _1: List(B), _2: List(C) | Type]
(vague A, vague B, vague C, implicit T, x) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | T]]]: (vague A: Type, vague B: Type, vague C: Type, implicit T: Type, x: T) -> [_: List(A), _1: List(B), _2: List(C) | Type]
(vague A, vague B, vague C, implicit T, x) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | T]]]: (vague A: Type, vague B: Type, vague C: Type, implicit T: Type, x: T) -> [_: List(A), _1: List(B), _2: List(C) | Type]
(vague A, vague B, vague C, vague D) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | List.null(vague D)]]]: (vague A: Type, vague B: Type, vague C: Type, vague D: Type) -> [_: List(A), _1: List(B), _2: List(C) | List(D)]
(vague A, vague B, vague C, vague D) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | List.null(vague D)]]]: (vague A: Type, vague B: Type, vague C: Type, vague D: Type) -> [_: List(A), _1: List(B), _2: List(C) | List(D)]
(vague A, vague B, vague C, vague D) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | List.null(vague D)]]]: (vague A: Type, vague B: Type, vague C: Type, vague D: Type) -> [_: List(A), _1: List(B), _2: List(C) | List(D)]
(vague A, vague B, vague C, vague D) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | List.null(vague D)]]]: (vague A: Type, vague B: Type, vague C: Type, vague D: Type) -> [_: List(A), _1: List(B), _2: List(C) | List(D)]
(vague A, vague B, vague C, vague D) => [List.null(vague A) | [List.null(vague B) | [List.null(vague C) | List.null(vague D)]]]: (vague A: Type, vague B: Type, vague C: Type, vague D: Type) -> [_: List(A), _1: List(B), _2: List(C) | List(D)]
